core/src/main/java/hivemall/ftvec/selection/ChiSquareUDF.java

```diff
@@ -0,0 +1,149 @@
+        }
+
+        if (!HiveUtils.isNumberListListOI(OIs[0])) {
+            throw new UDFArgumentTypeException(0,
+                "Only array<array<number>> type argument is acceptable but " + OIs[0].getTypeName()
+                        + " was passed as `observed`");
+        }
+
+        if (!HiveUtils.isNumberListListOI(OIs[1])) {
+            throw new UDFArgumentTypeException(1,
+                "Only array<array<number>> type argument is acceptable but " + OIs[1].getTypeName()
+                        + " was passed as `expected`");
+        }
+
+        observedOI = HiveUtils.asListOI(OIs[1]);
+        observedRowOI = HiveUtils.asListOI(observedOI.getListElementObjectInspector());
+        observedElOI = HiveUtils.asDoubleCompatibleOI(observedRowOI.getListElementObjectInspector());
+        expectedOI = HiveUtils.asListOI(OIs[0]);
+        expectedRowOI = HiveUtils.asListOI(expectedOI.getListElementObjectInspector());
+        expectedElOI = HiveUtils.asDoubleCompatibleOI(expectedRowOI.getListElementObjectInspector());
+
+        List<ObjectInspector> fieldOIs = new ArrayList<ObjectInspector>();
+        fieldOIs.add(ObjectInspectorFactory.getStandardListObjectInspector(PrimitiveObjectInspectorFactory.writableDoubleObjectInspector));
+        fieldOIs.add(ObjectInspectorFactory.getStandardListObjectInspector(PrimitiveObjectInspectorFactory.writableDoubleObjectInspector));
+
+        return ObjectInspectorFactory.getStandardStructObjectInspector(
+            Arrays.asList("chi2_vals", "p_vals"), fieldOIs);
+    }
+
+    @Override
+    public Object evaluate(GenericUDF.DeferredObject[] dObj) throws HiveException {
+        List observedObj = observedOI.getList(dObj[0].get()); // shape = (#classes, #features)
+        List expectedObj = expectedOI.getList(dObj[1].get()); // shape = (#classes, #features)
+
+        Preconditions.checkNotNull(observedObj);
+        Preconditions.checkNotNull(expectedObj);
+        final int nClasses = observedObj.size();
+        Preconditions.checkArgument(nClasses == expectedObj.size()); // same #rows
+
+        int nFeatures = -1;
+        double[] observedRow = null; // to reuse
+        double[] expectedRow = null; // to reuse
+        double[][] observed = null; // shape = (#features, #classes)
+        double[][] expected = null; // shape = (#features, #classes)
+
+        // explode and transpose matrix
+        for (int i = 0; i < nClasses; i++) {
+            if (i == 0) {
+                // init
+                observedRow = HiveUtils.asDoubleArray(observedObj.get(i), observedRowOI,
+                    observedElOI, false);
+                expectedRow = HiveUtils.asDoubleArray(expectedObj.get(i), expectedRowOI,
+                    expectedElOI, false);
+                nFeatures = observedRow.length;
+                observed = new double[nFeatures][nClasses];
+                expected = new double[nFeatures][nClasses];
```

Comment:
observed and expected SHOULD be instance variable and shared across `evaluate` calls.